<a href="https://colab.research.google.com/github/aserdargun/DSML101/blob/main/Part_1_Section_04_Numeric_Types.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **PART 1: FUNCTIONAL PROGRAMMING**

## Section 04 - Numeric Types

### 01 - Integers - Data Type

Integers are objects - instances of the `int` class

In [ ]:
print(type(100))

<class 'int'>


They are a variable length data type that can theoretically handle any integer magnitude. This will take up a variable amount of memory that depends on the particular size of the integer.

In [ ]:
import sys

Creating an integer object requires an overhead of 24 bytes.

In [ ]:
sys.getsizeof(0)

24

Here we see that to store the number 1 required 4 bytes (32 bits) on top of the 24 byte overhead:

In [ ]:
sys.getsizeof(1)

28

Larger numbers will require more storage space:

In [ ]:
sys.getsizeof(2**1000)

160

Larger integers will also slow down calcultaions.

In [ ]:
import time

In [ ]:
def calc(a):
    for i in range(10000000):
        a * 2

We start with a small integer value for a (10):

In [ ]:
start = time.perf_counter()
calc(10)
end = time.perf_counter()
print(end - start)

0.47732889999997496


Now we set a to something larger `2**100`:

In [ ]:
start = time.perf_counter()
calc(2**100)
end = time.perf_counter()
print(end - start)

1.0395105000000058


Finally we set a to some really large  value `2**10000`

In [ ]:
start = time.perf_counter()
calc(2**10000)
end = time.perf_counter()
print(end - start)

6.573301500000014


### 02 - Integers - Operations

Addition, subtraction, multiplication and exponentiation of integers always resul in an integer. (In the case of exponentiation this holds for positive integer exponents.)

In [ ]:
type(2 + 3)

int

In [ ]:
type(3 - 10)

int

In [ ]:
type(3 * 5)

int

In [ ]:
type(3 ** 4)

int

But the standard division operator `/` always results in a float value.

In [ ]:
type(2 / 3)

float

In [ ]:
type(10 / 2)

float

The `math.floor()` method will return the floor of any number.

In [ ]:
import math

For non-negative values (>=0), the floor of the value is the same as the integer portion of the value (truncation)

In [ ]:
math.floor(3.15)

3

In [ ]:
math.floor(3.9999999)

3

However, this is not the case for negative values:

In [ ]:
math.floor(-3.15)

-4

In [ ]:
math.floor(-3.0000001)

-4

**The Floor Division Operator**

The floor division operator `a//b` is the floor of `a / b`

i.e. `a // b = math.floor(a / b)`

This is true whether `a` and `b` are positive or negative.

In [ ]:
a = 33
b = 16
print(a/b)
print(a//b)
print(math.floor(a/b))

2.0625
2
2


For positive numbers, `a//b` is basically the same as truncating (taking the integer portion) of `a / b`.

But this is not the case for negative numbers.

In [ ]:
a = -33
b = 16
print('{0}/{1} = {2}'.format(a, b, a/b))
print('trunc({0}/{1}) = {2}'.format(a, b, math.trunc(a/b)))
print('{0}//{1} = {2}'.format(a, b, a//b))
print('floor({0}//{1}) = {2}'.format(a, b, math.floor(a/b)))

-33/16 = -2.0625
trunc(-33/16) = -2
-33//16 = -3
floor(-33//16) = -3


In [ ]:
a = 33
b = -16
print('{0}/{1} = {2}'.format(a, b, a/b))
print('trunc({0}/{1}) = {2}'.format(a, b, math.trunc(a/b)))
print('{0}//{1} = {2}'.format(a, b, a//b))
print('floor({0}//{1}) = {2}'.format(a, b, math.floor(a/b)))

33/-16 = -2.0625
trunc(33/-16) = -2
33//-16 = -3
floor(33//-16) = -3


**The Modulo Operator**
The modulo operator and the floor division operator will always satisfy the following equation:

`a = b * (a // b) + a % b`

In [ ]:
a = 13
b = 4
print('{0}/{1} = {2}'.format(a, b, a/b))
print('{0}//{1} = {2}'.format(a, b, a//b))
print('{0}%{1} = {2}'.format(a, b, a%b))
print(a == b * (a//b) + a%b)

13/4 = 3.25
13//4 = 3
13%4 = 1
True


In [ ]:
a = -13
b = 4
print('{0}/{1} = {2}'.format(a, b, a/b))
print('{0}//{1} = {2}'.format(a, b, a//b))
print('{0}%{1} = {2}'.format(a, b, a%b))
print(a == b * (a//b) + a%b)

-13/4 = -3.25
-13//4 = -4
-13%4 = 3
True


In [ ]:
a = 13
b = -4
print('{0}/{1} = {2}'.format(a, b, a/b))
print('{0}//{1} = {2}'.format(a, b, a//b))
print('{0}%{1} = {2}'.format(a, b, a%b))
print(a == b * (a//b) + a%b)

13/-4 = -3.25
13//-4 = -4
13%-4 = -3
True


In [ ]:
a = -13
b = -4
print('{0}/{1} = {2}'.format(a, b, a/b))
print('{0}//{1} = {2}'.format(a, b, a//b))
print('{0}%{1} = {2}'.format(a, b, a%b))
print(a == b * (a//b) + a%b)

-13/-4 = 3.25
-13//-4 = 3
-13%-4 = -1
True


### 03 - Integers - Constructors and Bases

**Constructors**
The `int` class has two constructors

In [ ]:
help(int)

Help on class int in module builtins:

class int(object)
 |  int([x]) -> integer
 |  int(x, base=10) -> integer
 |  
 |  Convert a number or string to an integer, or return 0 if no arguments
 |  are given.  If x is a number, return x.__int__().  For floating point
 |  numbers, this truncates towards zero.
 |  
 |  If x is not a number or if base is given, then x must be a string,
 |  bytes, or bytearray instance representing an integer literal in the
 |  given base.  The literal can be preceded by '+' or '-' and be surrounded
 |  by whitespace.  The base defaults to 10.  Valid bases are 0 and 2-36.
 |  Base 0 means to interpret the base from the string as an integer literal.
 |  >>> int('0b100', base=0)
 |  4
 |  
 |  Built-in subclasses:
 |      bool
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __bool__(self, /)
 |      True if 

In [ ]:
int(10)

10

In [ ]:
int(10.9)

10

In [ ]:
int(-10.9)

-10

In [ ]:
from fractions import Fraction

In [ ]:
a = Fraction(22, 7)

In [ ]:
a

Fraction(22, 7)

In [ ]:
int(a)

3

We can use the second constructor to generate integers (base 10) from strings in any base.

In [ ]:
int("10")

10

In [ ]:
int("101", 2)

5

In [ ]:
int("101", 2)

5

In [ ]:
int("101", base=2)

5

Python uses `a-z` for bases from 11 to 36.

---
**BE CAREFUL!**

*Note that the letters are not case sensitive.*

In [ ]:
int("F1A", base=16)

3866

int("f1a", base=16)

*Of course, the string must be a valid number in whatever base you specify.*

In [ ]:
int('B1A', base=11)

ValueError: invalid literal for int() with base 11: 'B1A'

In [ ]:
int('B1A', 12)

1606

**Base Representations**

**Built-ins**

In [ ]:
bin(10)

'0b1010'

In [ ]:
oct(10)

'0o12'

In [ ]:
hex(10)

'0xa'

---

**BE CAREFUL!**

*Note the `0b`, `0o`, `0x` prefixes.*

*You can use these in your own strings as well, and they correspond to prefixes used in integer literals as well.*

In [ ]:
a = int('1010', 2)
b = int('0b1010', 2)
c = 0b1010

In [ ]:
print(a, b, c)

10 10 10


In [ ]:
a = int('f1a', 16)
b = int('0xf1a', 16)
c = 0xf1a

In [ ]:
print(a, b, c)

3866 3866 3866


For literals, the `a-z` characters are not case-sensitive either.

In [ ]:
a = 0xf1a
b = 0xF1a
c = 0XF1A

In [ ]:
print(a, b, c)

3866 3866 3866


**Custom Rebasing**

Python only provides built-in function to rebase to base 2, 8, 16.

For other bases, you have to provide your own algorithm (or leverage some 3rd party library of your choice)

In [ ]:
def from_base10(n, b):
    if b < 2:
        raise ValueError('Base b must be >= 2')
    if n < 0:
        raise ValueError('Number n must be >= 0')
    if n == 0:
        return [0]
    digits = []
    while n > 0:
        # m = n % b
        # n = n // b
        # which is the same as:
        n, m = divmod(n, b)
        digits.insert(0, m)
    return digits

In [ ]:
from_base10(10, 2)

[1, 0, 1, 0]

In [ ]:
from_base10(255, 16)

[15, 15]

Next we may want to encode the digits into strings using different characters for each digit in the base.

In [ ]:
def encode(digits, digit_map):
    # we require that digit_map has at least as many
    # characters as the max number in digits
    if max(digits) >= len(digit_map):
        raise ValueError("digit_map is not long enough to encode digits")
    
    # we'll see this later, but the following would be better:
    encoding = ''.join([digit_map[d] for d in digits])
    return encoding

Now we can encode any list of digits:

In [ ]:
encode([1, 0, 1], "FT")

'TFT'

In [ ]:
encode([1, 10, 11], '0123456789AB')

'1AB'

And we can combine both functions into a sinhle one for easier use:

In [ ]:
def rebase_from10(number, base):
    digit_map = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'
    if base < 2 or base > 36:
        raise ValueError('Invalid base: 2 <= base <= 36')
    # we store the sign of number and maket it positive
    # we'll re-insert the sign at the end
    sign = -1 if number < 0 else 1
    number *= sign

    digits = from_base10(number, base)
    encoding = encode(digits, digit_map)
    if sign == -1:
        encoding = '-' + encoding
    return encoding

In [ ]:
e = rebase_from10(10, 2)
print(e)
print(int(e, 2))

1010
10


In [ ]:
e = rebase_from10(-10, 2)
print(e)
print(int(e, 2))

-1010
-10


In [ ]:
rebase_from10(131, 11)

'10A'

In [ ]:
rebase_from10(4095, 16)

'FFF'

In [ ]:
rebase_from10(-4095, 16)

'-FFF'

### 04 - Rational Numbers

In [ ]:
from fractions import Fraction

We can get some info on the Fraction class:

In [ ]:
help(Fraction)

Help on class Fraction in module fractions:

class Fraction(numbers.Rational)
 |  Fraction(numerator=0, denominator=None, *, _normalize=True)
 |  
 |  This class implements rational numbers.
 |  
 |  In the two-argument form of the constructor, Fraction(8, 6) will
 |  produce a rational number equivalent to 4/3. Both arguments must
 |  be Rational. The numerator defaults to 0 and the denominator
 |  defaults to 1 so that Fraction(3) == 3 and Fraction() == 0.
 |  
 |  Fractions can also be constructed from:
 |  
 |    - numeric strings similar to those accepted by the
 |      float constructor (for example, '-2.3' or '1e10')
 |  
 |    - strings of the form '123/456'
 |  
 |    - float and Decimal instances
 |  
 |    - other Rational instances (including integers)
 |  
 |  Method resolution order:
 |      Fraction
 |      numbers.Rational
 |      numbers.Real
 |      numbers.Complex
 |      numbers.Number
 |      builtins.object
 |  
 |  Methods defined here:
 |  
 |  __abs__(a)
 |    

We can create Fraction objects in a variety of ways:

Using integers:

In [ ]:
Fraction(1)

Fraction(1, 1)

In [ ]:
Fraction(1, 3)

Fraction(1, 3)

Using rational numbers:

In [ ]:
x = Fraction(2, 3)
y = Fraction(3, 4)
# 2/3 / 3/4 --> 2/3 * 4/3 --> 8/9
Fraction(x, y)

Fraction(8, 9)

Using floats:

In [ ]:
Fraction(0.125)

Fraction(1, 8)

In [ ]:
Fraction(0.5)

Fraction(1, 2)

Using strings:

In [ ]:
Fraction('10.5')

Fraction(21, 2)

In [ ]:
Fraction('22/7')

Fraction(22, 7)

Fractions are automatically reduced:

In [ ]:
Fraction(8, 16)

Fraction(1, 2)

Negative sign is attached to the numerator:

In [ ]:
Fraction(-1, 4)

Fraction(-1, 4)

Standard arithmetic operators are supported:

In [ ]:
Fraction(1, 3) + Fraction(1, 3) + Fraction(1, 3)

Fraction(1, 1)

In [ ]:
Fraction(1, 2) * Fraction(1, 4)

Fraction(1, 8)

In [ ]:
Fraction(1, 2) / Fraction(1, 3)

Fraction(3, 2)

We can recover the numerator and denominator (integers):

In [ ]:
x = Fraction(22, 7)
print(x.numerator)
print(x.denominator)

22
7


Since floats have finite precision, any float can be convertted to a rational number:

In [ ]:
import math
x = Fraction(math.pi)
print(x)
print(float(x))

884279719003555/281474976710656
3.141592653589793


In [ ]:
x = Fraction(math.sqrt(2))
print(x)

6369051672525773/4503599627370496


---

**BE CAREFUL!**

*Note that these rational values are approximations to the irrational number `pi` and `sqrt(2)`*

*Float number representations (as we will examine in future lessons) do not always have an exact representation.*

*The number 0.125 (1/8) has an exact representation:*

In [ ]:
Fraction(0.125)

Fraction(1, 8)

*and so we see the expected equivalent fraction.*

*But, 0.3 (3/10) does not have an exact representation:*

In [ ]:
Fraction(3, 10)

Fraction(3, 10)

*but*

In [ ]:
Fraction(0.3)

Fraction(5404319552844595, 18014398509481984)

We will study this in upcoming lessons.

Bur for now, let's just see a quick explanation:

In [ ]:
x = 0.3

In [ ]:
print(x)

0.3


Everything looks ok here - why am I saying 0.3 (float) is just an approximation?

Python is trying to format the displayed value for readability - so it rounds the number for a better display format!

We can instead choose to display the value using a certain number of digits:

In [ ]:
format(x, '.5f')

'0.30000'

At 5 digits after the decimal, we might still think 0.3 is an exact representation.

But let's display a few more digits:

In [ ]:
format(x, '.15f')

'0.300000000000000'

---
**BE CAREFUL!**

*Hmm... 15 digits and still looking good!*

*How about 25 digits...*

In [ ]:
format(x, '.25f')

'0.2999999999999999888977698'

Now we see that `x` is not quite 0.3...

In fact, we can quantify the delta this way:

In [ ]:
delta = Fraction(0.3) - Fraction(3, 10)

Theoretically, delta should be 0, but it's not:

In [ ]:
delta == 0

False

In [ ]:
delta

Fraction(-1, 90071992547409920)

`delta` is a very small number, the above fraction...

As a float:

In [ ]:
float(delta)

-1.1102230246251566e-17

**Constraining the denominator**

In [ ]:
x = Fraction(math.pi)
print(x)
print(format(float(x), '.25f'))

884279719003555/281474976710656
3.1415926535897931159979635


In [ ]:
y = x.limit_denominator(10)
print(y)
print(format(float(y), '.25f'))

22/7
3.1428571428571427937015414


In [ ]:
y = x.limit_denominator(100)
print(y)
print(format(float(y), '.25f'))

311/99
3.1414141414141414365701621


In [ ]:
y = x.limit_denominator(500)
print(y)
print(format(float(y), '.25f'))

355/113
3.1415929203539825209645642


### 05 - Floats - Internal Representation

The `float` class can be used to represent real numbers.

In [ ]:
help(float)

Help on class float in module builtins:

class float(object)
 |  float(x=0, /)
 |  
 |  Convert a string or number to a floating point number, if possible.
 |  
 |  Methods defined here:
 |  
 |  __abs__(self, /)
 |      abs(self)
 |  
 |  __add__(self, value, /)
 |      Return self+value.
 |  
 |  __bool__(self, /)
 |      True if self else False
 |  
 |  __ceil__(self, /)
 |      Return the ceiling as an Integral.
 |  
 |  __divmod__(self, value, /)
 |      Return divmod(self, value).
 |  
 |  __eq__(self, value, /)
 |      Return self==value.
 |  
 |  __float__(self, /)
 |      float(self)
 |  
 |  __floor__(self, /)
 |      Return the floor as an Integral.
 |  
 |  __floordiv__(self, value, /)
 |      Return self//value.
 |  
 |  __format__(self, format_spec, /)
 |      Formats the float according to format_spec.
 |  
 |  __ge__(self, value, /)
 |      Return self>=value.
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __getnewargs__(self, /)
 

The `float` class has a single constructor, which can take a number or a string and will attempt to convert it to a float.

In [ ]:
float(10)

10.0

In [ ]:
float(3.14)

3.14

In [ ]:
float('0.1')

0.1

---

**BE CAREFUL**

*However, strings that represent fractions cannot be converted to floats, unlike the Fraction class we saw earlier.*

In [ ]:
float('22/7')

ValueError: could not convert string to float: '22/7'

*If you really want to get a float from a string such as `'22/7'`, you could first create a `Fraction`, then create a `float` from that:*

In [ ]:
from fractions import Fraction

In [ ]:
float(Fraction('22/7'))

3.142857142857143

*Floats do not always have an exact representation:*

In [ ]:
print(0.1)

0.1


*Although this looks like `0.1` exactly, we need to reveal more digits after the decimal point to see what's going on:*

In [ ]:
format(0.1, '.25f')

'0.1000000000000000055511151'

*However, certain numbers can be represented exactly in a binary fraction expansion:*

In [ ]:
format(0.125, '.25f')

'0.1250000000000000000000000'

*This is because 0.125 is precisely 1/8 or 1/(2^3)*

### 06 - Floats - Equality Testing

Because not all real numbers have an exact `float` representation, equality testing can be tricky.

In [ ]:
x = 0.1 + 0.1 + 0.1
y = 0.3
x == y

False

This is because `0.1` and `0.3` do not have exact representations:

In [ ]:
print('0.1 --> {0:.25f}'.format(0.1))
print('x --> {0:.25f}'.format(x))
print('y --> {0:.25f}'.format(y))

0.1 --> 0.1000000000000000055511151
x --> 0.3000000000000000444089210
y --> 0.2999999999999999888977698


However, in some (limited) cases where all the numbers involved do have an exact representation, it will work:

In [ ]:
x = 0.125 + 0.125 + 0.125
y = 0.375
x == y

True

In [ ]:
print('0.125 --> {0:.25f}'.format(0.125))
print('x --> {0:.25f}'.format(x))
print('y --> {0:.25f}'.format(y))

0.125 --> 0.1250000000000000000000000
x --> 0.3750000000000000000000000
y --> 0.3750000000000000000000000


One simple way to get around this is to round to a specific number of digits and then compare.

In [ ]:
x = 0.1 + 0.1 + 0.1
y = 0.3
round(x, 5) == round(y, 5)

True

We can also use a more flexible technique implemented by the `isclose` method in the `math` module.

In [ ]:
from math import isclose

In [ ]:
help(isclose)

Help on built-in function isclose in module math:

isclose(a, b, *, rel_tol=1e-09, abs_tol=0.0)
    Determine whether two floating point numbers are close in value.
    
      rel_tol
        maximum difference for being considered "close", relative to the
        magnitude of the input values
      abs_tol
        maximum difference for being considered "close", regardless of the
        magnitude of the input values
    
    Return True if a is close in value to b, and False otherwise.
    
    For the values to be considered close, the difference between them
    must be smaller than at least one of the tolerances.
    
    -inf, inf and NaN behave similarly to the IEEE 754 Standard.  That
    is, NaN is not close to anything, even itself.  inf and -inf are
    only close to themselves.



In [ ]:
x = 0.1 + 0.1 + 0.1
y = 0.3
isclose(x, y)

True

The `isclose` method takes two optional parameters, `rel_tol` and `abs_tol`.

`rel_tol` is a relative tolerance that will be relative to the magnitude of the largest of the two numbers being compared. Useful when we want to see if two numbers are close to each other as a percentage of their magnitudes.

`abs_tol` is an absolute tolerance that is independent of the magnitunde of the numbers we are compairing - this is useful for numbers that are close to zero.

In this situation we might consider x and y to be close to each other:

In [ ]:
x = 123456789.01
y = 123456780.02

but not in this case:

In [ ]:
x = 0.01
y = 0.02

In both these cases the difference between the two numbers was `0.01`, yet in one case we considered the numvers "equal" and in the other, not "equal". Realtive tomerances are useful to handle thse scenarios.

In [ ]:
isclose(123456789.01, 123456789.02, rel_tol=0.01)

True

In [ ]:
isclose(0.01, 0.02, rel_tol=0.01)

False

On the other hand, we have to be careful with relative tolerances when working with values that are close to zero:

In [ ]:
x = 0.0000001
y = 0.0000002
isclose(x, y, rel_tol=0.01)

False

So, we could use an absolute tolerance here:

In [ ]:
isclose(x, y, abs_tol=0.0001, rel_tol=0)

True

In general, we can combine the use of both relative and absolute tolerances in this way:

In [ ]:
x = 0.0000001
y = 0.0000002

a = 123456789.01
b = 123456789.02

print('x = y:', isclose(x, y, abs_tol=0.0001, rel_tol=0.01))
print('a = b:', isclose(a, b, abs_tol=0.0001, rel_tol=0.01))

x = y: True
a = b: True


### 07 - Floats - Coercing to Integers

**Truncation**

In [ ]:
from math import trunc

In [ ]:
trunc(10.3), trunc(10.5), trunc(10.6)

(10, 10, 10)

In [ ]:
trunc(-10.6), trunc(-10.5), trunc(-10.3)

(-10, -10, -10)

The `ìnt` constructor uses truncation when a float is passed in:

In [ ]:
int(10.3), int(10.5), int(10.6)

(10, 10, 10)

In [ ]:
int(-10.5), int(-10.5), int(-10.4)

(-10, -10, -10)

**Floor**

In [ ]:
from math import floor

In [ ]:
floor(10.4), floor(10.5), floor(10.6)

(10, 10, 10)

In [ ]:
floor(-10.4), floor(-10.5), floor(-10.6)

(-11, -11, -11)

**Ceiling**

In [ ]:
from math import ceil

In [ ]:
ceil(10.4), ceil(10.5), ceil(10.6)

(11, 11, 11)

In [ ]:
ceil(-10.4), ceil(-10.5), ceil(-10.6)

(-10, -10, -10)

### 08 - Floats - Rounding

In [ ]:
help(round)

Help on built-in function round in module builtins:

round(number, ndigits=None)
    Round a number to a given precision in decimal digits.
    
    The return value is an integer if ndigits is omitted or None.  Otherwise
    the return value has the same type as the number.  ndigits may be negative.



**n = 0**

In [ ]:
a = round(1.5)
a, type(a)

(2, int)

In [ ]:
a = round(1.5, 0)
a, type(a)

(2.0, float)

**n > 0**

In [ ]:
round(1.8888, 3), round(1.8888, 2), round(1.8888, 1), round(1.8888, 0)

(1.889, 1.89, 1.9, 2.0)

**n < 0**

In [ ]:
round(888.88, 1), round(888.88, 0), \
round(888.88, -1), round(888.88, -2), \
round(888.88, -3)

(888.9, 889.0, 890.0, 900.0, 1000.0)

**Ties**

In [ ]:
round(1.25, 1)

1.2

In [ ]:
round(1.35, 1)

1.4

This is rounding to nearest, with ties to nearest number with even least significant digit, ake Banker's Rounding.

Works similarly with n negative.

In [ ]:
round(15, -1)

20

In [ ]:
round(25, -1)

20

---

**BE CAREFUL!**

**Rounding to closest, ties away from zero**

*This is traditionally the type of rounding taught in school, which is different from the Banker's Rounding implemented in Python (and in many other programming languages)*

*This is rounding to nearest even number.*

1.5 --> 2

2.5 --> 3

-1.5 --> -2
-2.5 --> -3

To do this thepe of rounding (to nearest 1) we can add (for positive numbers) or subtract (for negative numbers) 0.5 and then truncate the resulting number.

In [ ]:
def _round(x):
    from math import copysign
    return int(x + 0.5 * copysign(1, x))

In [ ]:
round(1.5), _round(1.5)

(2, 2)

In [ ]:
round(2.5), _round(2.5)

(2, 3)

### 09 - Decimals

In [ ]:
import decimal

In [ ]:
from decimal import Decimal

Decimals have context, that can be used to specify rounding and precision (amongst other things)

Contexts can be local (temporary contexts) or global (default)

**Global Context**

In [ ]:
g_ctx = decimal.getcontext()

In [ ]:
g_ctx.prec

28

In [ ]:
g_ctx.prec = 6

In [ ]:
g_ctx.rounding = decimal.ROUND_HALF_UP

And if we read this back directly from the global context:

In [ ]:
decimal.getcontext().prec

6

In [ ]:
decimal.getcontext().rounding

'ROUND_HALF_UP'

we see that the global context was indeed changed.

**Local Context**

The `localcontext()` function will return a context manager thar we can use with a `with` statement:

In [ ]:
with decimal.localcontext() as ctx:
    print(ctx.prec)
    print(ctx.rounding)

6
ROUND_HALF_UP


Since no argument was specified in the `localcontext()` call, it provides us a context manager that uses a copy of the global context.

Modifying the local context has no effect on the global context.

In [ ]:
with decimal.localcontext() as ctx:
    ctx.prec = 10
    print('local prec = {0}, global prec = {1}'.format(ctx.prec, g_ctx.prec))

local prec = 10, global prec = 6


**Rounding**

In [ ]:
decimal.getcontext().rounding

'ROUND_HALF_UP'

The rounding mechanism is ROUND_HALF_UP because we set the global context to that earlier in this notebook. Note that normally the default is ROUND_HALF_EVEN.

So we first reset our global context rounding to that:

In [ ]:
decimal.getcontext().rounding = decimal.ROUND_HALF_EVEN

In [ ]:
x = Decimal('1.25')
y = Decimal('1.35')
print(round(x, 1))
print(round(y, 1))

1.2
1.4


Let's change the rounding mechanism in the global context to ROUND_HALF_UP

In [ ]:
decimal.getcontext().rounding = decimal.ROUND_HALF_UP

In [ ]:
x = Decimal('1.25')
y = Decimal('1.35')
print(round(x, 1))
print(round(y, 1))

1.3
1.4


As you may have realized, changing the global context is a pain if you need to constantly switch between precisions and rounding algorithms. Also, it could introduce bugs if you forget that you changed the global context somewhere further up in your module.

For this reason, it is usually better to use a local context manager instead:

First we reset our global context rounding to the default:

In [ ]:
decimal.getcontext().rounding = decimal.ROUND_HALF_EVEN

In [ ]:
from string import ascii_letters
x = Decimal('1.25')
y = Decimal('1.35')
print(round(x, 1), round(y, 1))
with decimal.localcontext() as ctx:
    ctx.rounding = decimal.ROUND_HALF_UP
    print(round(x, 1), round(y, 1))
print(round(x, 1), round(y, 1))

1.2 1.4
1.3 1.4
1.2 1.4


### 10 - Decimals - Constructors and Contexts

The `Decimal` constructor can handle a variety of data types.

In [ ]:
import decimal
from decimal import Decimal

**Integers**

In [ ]:
Decimal(10)

Decimal('10')

In [ ]:
Decimal(-10)

Decimal('-10')

**Strings**

In [ ]:
Decimal('0.1')

Decimal('0.1')

In [ ]:
Decimal('-3.1415')

Decimal('-3.1415')

**Tuples**

In [ ]:
Decimal((0, (3,1,4,1,4), -4))

Decimal('3.1414')

In [ ]:
Decimal((1, (1,2,3,4), -3))

Decimal('-1.234')

In [ ]:
Decimal((0, (1,2,3), 3))

Decimal('1.23E+5')

---

**BE CAREFUL!**

*Tuple components are sign=0, digits=(3,1,4,1,4), exponent=-4 but we don't use parameters with this key=value format.*

`Decimal((0, (3,1,4,1,4), -4))`

---

**BE CAREFUL!**

**But don't use Floats**

In [ ]:
format(0.1, '.25f')

'0.1000000000000000055511151'

In [ ]:
Decimal(0.1)

Decimal('0.1000000000000000055511151231257827021181583404541015625')

As you can see, since we passed an approximate binary float to the Decimal constructor it did it's best to represent that binary float exactly!!

So, instead, use strings or tuples in the Decimal constructor.

**Context Precision and the Constructor**

The context precision does not affect the precision used when creating a Decimal object - those are independent of each other.

Let's set our global (default) context to a precision of 2

In [ ]:
decimal.getcontext().prec = 2

Now we can create decimal numbers of higher precision than that:

In [ ]:
a = Decimal('0.12345')
b = Decimal('0.12345')

In [ ]:
a

Decimal('0.12345')

In [ ]:
b

Decimal('0.12345')

But when we add those two numbers up, the context precision will matter:

In [ ]:
decimal.getcontext().prec = 2
a + b

Decimal('0.25')

As you can see, we ended up with a sum that was rounded to 2 digits after the decimal point (precision = 2)

**Local and Global Contexts are Independent**

In [ ]:
decimal.getcontext().prec = 6

In [ ]:
decimal.getcontext().rounding

'ROUND_HALF_EVEN'

In [ ]:
a = Decimal('0.12345')
b = Decimal('0.12345')
print(a + b)
with decimal.localcontext() as ctx:
    ctx.prec = 2
    c = a + b
    print('c within local context: {0}'.format(c))
print('c within global context: {0}'.format(c))

0.24690
c within local context: 0.25
c within global context: 0.25


Since c was created within the local context by adding a and b, and the local context had a precision of 2, c was rounded to 2 digits after the decimal point.

Once the local context is destroyed (after the with block), the variable c still exists, and its precision is still just 2 - it doesn't magically suddenly get the global context's precision of 6.

### 11 - Decimals - Math Operations

**Div and Mod**

The `//` and `%` operators (and consequently, the `divmod()` function) behave differently for integers and Decimals.

Thi is because integer division for Decimals is performed differently, and results in a truncated division, whereas integers use a floored division.

These differences are only when negative number are involved. If all numbers involved are positive, then integer and Decimal div and mod operations are equal.

But in both cases the `//` and `%` operators satisfy the equation:

`n = d * (n //d ) + (n % d)`

In [ ]:
import decimal
from decimal import Decimal

In [ ]:
x = 10
y = 3
print(x//y, x%y)
print(divmod(x, y))
print( x ==  y * (x//y) + x % y)

3 1
(3, 1)
True


In [ ]:
a = Decimal('10')
b = Decimal('3')
print(a//b, a%b)
print(divmod(a,b))
print( a == b * (a//b) + a % b)

3 1
(Decimal('3'), Decimal('1'))
True


As we can see, the `//` and `%` operators had the same result when both numbers were positive.

In [ ]:
x = -10
y = 3
print(x//y, x%y)
print(divmod(x, y))
print( x == y * (x//y) + x % y)

-4 2
(-4, 2)
True


In [ ]:
a = Decimal('-10')
b = Decimal('3')
print(a//b, a%b)
print(divmod(a, b))
print( a == b * (a//b) + a % b)

-3 -1
(Decimal('-3'), Decimal('-1'))
True


On the other hand, we see that in this case the `//` and `% ` operators did not result in the same values, although the equation was satisfied in both instances.

**Other Mathematical Functions**

The Decimal class implements a variety of mathematical functions.

In [ ]:
a = Decimal('1.5')
print(a.log10()) # base 10 logarithm
print(a.ln())    # natural logarithm (base e)
print(a.exp())   # e ** a
print(a.sqrt())  # square root

0.176091
0.405465
4.48169
1.22474


Although you can use the math function of the math module, be aware that the math module functions will cast the Decimal numbers to floats when it performs the various operations. So, if the precision is important (which it probably is if you decided to use Decimal numbers in the first place), choouse the math functions of the Decimal class over those of the math module.

In [ ]:
x = 2
x_dec = Decimal(2)

In [ ]:
import math

In [ ]:
root_float = math.sqrt(x)
root_mixed = math.sqrt(x_dec)
root_dec = x_dec.sqrt()

In [ ]:
print(format(root_float, '1.27f'))
print(format(root_mixed, '1.27f'))
print(root_dec)

1.414213562373095145474621859
1.414213562373095145474621859
1.41421


In [ ]:
print(format(root_float * root_float, '1.27f'))
print(format(root_mixed * root_mixed, '1.27f'))
print(root_dec * root_dec)

2.000000000000000444089209850
2.000000000000000444089209850
1.99999


In [ ]:
x = 0.01
x_dec = Decimal('0.01')

root_float = math.sqrt(x)
root_mixed = math.sqrt(x_dec)
root_dec = x_dec.sqrt()

print(format(root_float, '1.27f'))
print(format(root_mixed, '1.27f'))
print(root_dec)

0.100000000000000005551115123
0.100000000000000005551115123
0.1


In [ ]:
print(format(root_float * root_float, '1.27f'))
print(format(root_mixed * root_mixed, '1.27f'))
print(root_dec * root_dec)

0.010000000000000001942890293
0.010000000000000001942890293
0.01


### 12 - Decimals - Performance Considerations

**Memory Footprint**

Decimals take up a lot more memory than floats.

In [ ]:
import sys
from decimal import Decimal

In [ ]:
a = 3.1415
b = Decimal('3.1415')

In [ ]:
sys.getsizeof(a)

24

24 bytes are used to store the gloat 3.1415

In [ ]:
sys.getsizeof(b)

104

104 bytes are used to store the Decimal 3.1415

**Computational Performance**

Decimal arithmetic is also much slower than float arithmetic (on a CPUi an even more so if using a GPU)

We can do some rough timings to illustrate this.

First we look at the performance difference creating floats vs decimals:

In [ ]:
import time
from decimal import Decimal

def run_float(n=1):
    for i in range(n):
        a = 3.1415

def run_decimal(n=1):
    for i in range(n):
        a = Decimal('3.1415')

Timing float and Decimal operations:

In [ ]:
n = 10000000

In [ ]:
start = time.perf_counter()
run_float(n)
end = time.perf_counter()
print('float: ', end-start)

start = time.perf_counter()
run_decimal(n)
end = time.perf_counter()
print('decimal: ', end-start)

float:  0.36456689999977243
decimal:  5.793422599999758


We make a slight variant here to see how addition compares between the two types:

In [ ]:
def run_float(n=1):
    a = 3.1415
    for i in range(n):
        a + a
    
def run_decimal(n=1):
    a = Decimal('3.1415')
    for i in range(n):
        a + a

start = time.perf_counter()
run_float(n)
end = time.perf_counter()
print('float: ', end-start)

start = time.perf_counter()
run_decimal(n)
end = time.perf_counter()
print('decimal: ', end-start)

float:  0.8951041000000259
decimal:  1.6607788999999684


How about square roots:

(We drop the n count a bit)

In [ ]:
n = 5000000

import math

def run_float(n=1):
    a = 3.1415
    for i in range(n):
        math.sqrt(a)
        
def run_decimal(n=1):
    a = Decimal('3.1415')
    for i in range(n):
        a.sqrt()
        
start = time.perf_counter()
run_float(n)
end = time.perf_counter()
print('float: ', end-start)

start = time.perf_counter()
run_decimal(n)
end = time.perf_counter()
print('decimal: ', end-start)

float:  1.1838782000004358
decimal:  7.016659200000504


### 13 - Complex Numbers

Python's built-in class provides support for complex numbers.

Complex numbers are defined in rectangular coordinates (real and imaginary parts) using either the constructor or a literal expression.

The complex number `1 + 2j` can be defined in either of these ways:

In [ ]:
a = complex(1, 2)
b = 1 + 2j

In [ ]:
a == b

True

Note that the real and imaginary parts are defined as floats, and can be retrieved as follows:

In [ ]:
a.real, type(a.real)

(1.0, float)

In [ ]:
a.imag, type(a.imag)

(2.0, float)

The complex conjugate can be calculated as follows:

In [ ]:
a.conjugate()

(1-2j)

The standard arithmetic operators are polymorphic and defined for complex numbers.

In [ ]:
a = 1 + 2j
b = 3 - 4j
c = 5j
d = 10

In [ ]:
a + b

(4-2j)

In [ ]:
b * c

(20+15j)

In [ ]:
c / d

0.5j

In [ ]:
d - a

(9-2j)

The `//` and `%` operators, although also polymorphic, are not defined for complex numbers:

In [ ]:
a // b

TypeError: can't take floor of complex number.

In [ ]:
a % b

TypeError: can't mod complex numbers.

The `==` and `!=` operators support complex numbers - but since the real and imaginary parts of complex numbers are floats, the same problems comparing floats using `==` and `!=` also apply to complex numbers.

In [ ]:
a = 0.1j

In [ ]:
a + a + a == 0.3j

False

In addition, the standard comparison operators `(<, <=, >, >=)` are not defined for complex numbers.

In [ ]:
a = 1 + 1j
b = 100 + 100j
a < b

TypeError: '<' not supported between instances of 'complex' and 'complex'

**Math Functions**

The `cmath` module provides complex alternatives to the standard `math` functions.

In addition, the `cmath` module provides the complex implementation of the `isclose()` method available for floats.

In [ ]:
import cmath

a = 1 + 5j
print(cmath.sqrt(a))

(1.7462845577958914+1.4316108957382214j)


The standard `math` module functions will not work with complex numbers:

In [ ]:
import math
print(math.sqrt(a))

TypeError: can't convert complex to float

**Polar / Rectangular Conversions**

The `cmath.phase()` function can be used to return the phase (or argument) of any complex number.

The standard `abs()` function supports complex numbers and will return the magnitude (euclidean norm) of the complex number.

In [ ]:
a = 1 + 1j

In [ ]:
r = abs(a)
phi = cmath.phase(a)
print('{0} = ({1},{2})'.format(a, r, phi))

(1+1j) = (1.4142135623730951,0.7853981633974483)


Complex numbers in polar coordinates can be converted to rectangular coordinates using the `math.rect()` function:

In [ ]:
r = math.sqrt(2)
phi = cmath.pi/4
print(cmath.rect(r, phi))

(1.0000000000000002+1.0000000000000002j)


**Euler's Identity and the isclose() function**

`e^i*pi + 1 = 0`

In [ ]:
RHS = cmath.exp(cmath.pi * 1j) + 1
print(RHS)

1.2246467991473532e-16j


Which, because of limited precision is not quite zero.

However, the result is very close to zero.

We can use the `isclose()` method of the `cmath` module, which behaves similarly to the `math.isclose()` method. Since we are testing for closeness of two numbers close to zero, we need to make sure an absolute tolerance is also specified:

In [ ]:
cmath.isclose(RHS, 0, abs_tol=0.00001)

True

If we had not specified an absolute tolerance:

In [ ]:
cmath.isclose(RHS, 0)

False

### 14 - Booleans

The `bool` class is used to represent boolean values.

The `bool` class inherits from the `int` class.

In [ ]:
issubclass(bool, int)

True

Two built-in constants, `True` and `False` are singleton instances of the bool class with underlying int varues of 1 and 0 respectively.

In [ ]:
type(True), id(True), int(True)

(bool, 140735898179688, 1)

In [ ]:
type(False), id(False), int(False)

(bool, 140735898179720, 0)

These two values are instances of the `bool` class, and by inheritance are also `int` objects.

In [ ]:
isinstance(True, bool)

True

In [ ]:
isinstance(True, int)

True

Since `True` and `False` are singletons, we can use either the `is` operator, or the `==` operator to compare them to any boolean expression.

In [ ]:
id(True), id(1 < 2)

(140735898179688, 140735898179688)

In [ ]:
id(False), id(1 == 3)

(140735898179720, 140735898179720)

In [ ]:
(1 < 2) is True, (1 < 2) == True

(True, True)

In [ ]:
(1 == 2) is False, (1 == 2) == False

(True, True)

---
**BE CAREFUL!**

*Be careful with that last comparsion, the parentheses are necessary!*

In [ ]:
1 == 2 == False

False

In [ ]:
(1 == 2) == False

True

*We'll look into this in detail later, but, for now, this happens because a chained comparison such as `a == b == c` is actually evaluated as `a == b and b == c`*

*So `1 == 2 == False` is the same as `1 == 2 and 2 == False`*

In [ ]:
1 == 2, 2 == False, 1==2 and 2==False

(False, False, False)

*But,*

In [ ]:
(1 == 2)

False

*So ``(1 == 2) == False` evaluates to True*

*But since `False` is also `0`, we get the following:*

In [ ]:
(1 == 2) == 0

True

The underlying integer values of True and False are:

In [ ]:
int(True), int(False)

(1, 0)

So, using an equality comparison:

In [ ]:
1 == True, 0 == False

(True, True)

But, from an object perspective 1 and True are not the same (similarly with 0 and False)

In [ ]:
1 == True, 1 is True

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
C:\Users\aserd\AppData\Local\Temp\ipykernel_2072\1072280335.py:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  1 == True, 1 is True


(True, False)

In [ ]:
0 == False, 0 is False

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
C:\Users\aserd\AppData\Local\Temp\ipykernel_2072\2948352474.py:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  0 == False, 0 is False


(True, False)

Any integer can be cast to a boolean, and follows the rule:

bool(x) = True for any x except for zero which returns False

In [ ]:
bool(0)

False

In [ ]:
bool(1), bool(100), bool(-1)

(True, True, True)

Since booleans are subclassed from integers, they can behave like integers, and because of polymorphism all the standard integer operators, properties and methods apply.

In [ ]:
True > False

True

In [ ]:
True + 2

3

In [ ]:
False // 2

0

In [ ]:
True + True + True

3

In [ ]:
(True + True + True) % 2

1

In [ ]:
-True

-1

In [ ]:
100 * False

0

I certainly do not recommend you write code like that shown above, but aware that it does work.

### 15 - Booleans - Truth Values

All objects in Python have an associated `truth value`, or `truthyness`

We saw in a previous lecture that integers have an inherent truth value:

In [ ]:
bool(0)

False

In [ ]:
bool(1), bool(-1), bool(100)

(True, True, True)

This truthyness has nothing to do with the fact that `bool` is a subclass of `int`.

Instead, it has to do with the fact that the `int` class implements a `__bool__()` method:

In [ ]:
help(bool)

Help on class bool in module builtins:

class bool(int)
 |  bool(x) -> bool
 |  
 |  Returns True when the argument x is true, False otherwise.
 |  The builtins True and False are the only two instances of the class bool.
 |  The class bool is a subclass of the class int, and cannot be subclassed.
 |  
 |  Method resolution order:
 |      bool
 |      int
 |      object
 |  
 |  Methods defined here:
 |  
 |  __and__(self, value, /)
 |      Return self&value.
 |  
 |  __or__(self, value, /)
 |      Return self|value.
 |  
 |  __rand__(self, value, /)
 |      Return value&self.
 |  
 |  __repr__(self, /)
 |      Return repr(self).
 |  
 |  __ror__(self, value, /)
 |      Return value|self.
 |  
 |  __rxor__(self, value, /)
 |      Return value^self.
 |  
 |  __xor__(self, value, /)
 |      Return self^value.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create 

If you scroll down in the documentation you should reach a section that looks like this:

`| __bool__(self, /)`

`|    self != 0`

So, when we write:

In [ ]:
bool(100)

True

In [ ]:
(100).__bool__()

True

In [ ]:
(0).__bool__()

False

Most objects will implement either the `__bool__()` or `__len__()` methods. If they don't, then their associated value will be `True` always.

**Numeric Types**

Any non-zero numeric value is truthy. Any zero numeric value is falsy:

In [ ]:
from fractions import Fraction
from decimal import Decimal
bool(10), bool(1.5), bool(Fraction(3, 4)), bool(Decimal('10.5'))

(True, True, True, True)

In [ ]:
bool(0), bool(0.0), bool(Fraction(0,1)), bool(Decimal('0')), bool(0j)

(False, False, False, False, False)

**Sequence Types**

An empty sequence type object is Falsy, a non-empty one is truthy:

In [ ]:
bool([1, 2, 3]), bool((1, 2, 3)), bool('abc'), bool(1j)

(True, True, True, True)

In [ ]:
bool([]), bool(()), bool('')

(False, False, False)

**Mapping Types**

Similarly, an empty mapping type will be falsy, a non-empty one truthy:

In [ ]:
bool({'a': 1}), bool({1, 2, 3})

(True, True)

In [ ]:
bool({}), bool(set())

(False, False)

**The None Object**

The Singleton `None` object is always falsy:

In [ ]:
bool(None)

False

**One Application of Truth Values**

Any conditional expression which involves objects other than `bool` types, will use the associated truth value as the result of the conditional expression.

In [ ]:
a = [1, 2, 3]
if a:
    print(a[0])
else:
    print('a is None, or a is empty')

1


In [ ]:
a = []
if a:
    print(a[0])
else:
    print('a is None, or a is empty')

a is None, or a is empty


In [ ]:
a = 'abc'
if a:
    print(a[0])
else:
    print('a is None, or a is empty')

a


In [ ]:
a = ''
if a:
    print(a[0])
else:
    print('a is None, or a is empty')

a is None, or a is empty


We could write this using a more lengthy expression:

In [ ]:
a = 'abc'
if a is not None and len(a) > 0:
    print(a[0])
else:
    print('a is None, or a is empty')

a


Doing the following would break our code in some instances:

In [ ]:
a = 'abc'
if a is not None:
    print(a[0])

a


works, but:

In [ ]:
a = ''
if a is not None:
    print(a[0])

IndexError: string index out of range

In [ ]:
a = None
if len(a) > 0:
    print(a[0])

TypeError: object of type 'NoneType' has no len()

to be through we would need to write:

In [ ]:
a = None
if a is not None and len(a) > 0:
    print(a[0])

Also, the order of the boolean expressions matter here!

We'll discuss this and short-circuit evaluations.

For example:

In [ ]:
a = None
if len(a) > 0 and a is not None:
    print(a[0])

TypeError: object of type 'NoneType' has no len()

### 16 - Booleans - Precedence and Short-Circuiting

In [ ]:
True or True and False

True

this is equivalent, becaouse of `and` having higher precedence than `or`, to:

In [ ]:
True or (True and False)

True

This is not the same as:

In [ ]:
(True or True) and False

False

**Short-Circuiting**

In [ ]:
a = 100
b = 0

if a/b > 2:
    print('a is at least double b')

ZeroDivisionError: division by zero

In [ ]:
a = 10
b = 0
if b and a/b > 2:
    print('a is at least double b')

Can also be useful to deal with null or empty strings in a database:

In [ ]:
import string

In [ ]:
help(string)

Help on module string:

NAME
    string - A collection of string constants.

MODULE REFERENCE
    https://docs.python.org/3.9/library/string
    
    The following documentation is automatically generated from the Python
    source files.  It may be incomplete, incorrect or include features that
    are considered implementation detail and may vary between Python
    implementations.  When in doubt, consult the module reference at the
    location listed above.

DESCRIPTION
    Public module variables:
    
    whitespace -- a string containing all ASCII whitespace
    ascii_lowercase -- a string containing all ASCII lowercase letters
    ascii_uppercase -- a string containing all ASCII uppercase letters
    ascii_letters -- a string containing all ASCII letters
    digits -- a string containing all ASCII decimal digits
    hexdigits -- a string containing all ASCII hexadecimal digits
    octdigits -- a string containing all ASCII octal digits
    punctuation -- a string containing all

In [ ]:
string.digits

'0123456789'

In [ ]:
string.ascii_letters

'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'

In [ ]:
name = ''
if name[0] in string.digits:
    print('Name cannot start with a digit!')

IndexError: string index out of range

In [ ]:
name = ''
if name and name[0] in string.digits:
    print('Name cannot start with a digit!')

In [ ]:
name = None
if name and name[0] in string.digits:
    print('Name cannot start with a digit!')

In [ ]:
name = 'Bob'
if name and name[0] in string.digits:
    print('Name cannot start with a digit!')

In [ ]:
name = '1Bob'
if name and name[0] in string.digits:
    print('Name cannot start with a digit!')

Name cannot start with a digit!


### 17 - Booleans: Boolean Operators

The way the Boolean operators `and`, `or` actually work is a little different in Python:

**or**

`X or Y`: if X is falsy, return Y, otherwise evaluates and returns X

In [ ]:
'' or 'abc'

'abc'

In [ ]:
0 or 100

100

In [ ]:
[] or [1, 2, 3]

[1, 2, 3]

In [ ]:
[1, 2] or [1, 2, 3]

[1, 2]

---

**BE CAREFUL!**

*You should note that the truth value of `Y` is never even considered when evaluating the `or` result.*

*Only the left operand matters.*

*Of course, Y will be evaluated if it is being returned - bur its truth value does not affect how the `or` is being calculated.*

*You probably will notice that this means `Y` is not evaluated if `X` is returned - short-circuiting!!*

*We could(almost!) write the `or` operator ourselves in this way:*

In [ ]:
def _or(x, y):
    if x:
        return x
    else:
        return y

In [ ]:
print(_or(0, 100) == (0 or 100))
print(_or(None, 'n/a') == (None or 'n/a'))
print(_or('abc', 'n/a') == ('abc' or 'n/a'))

True
True
True


Why did I say almost?

Unlike the `or` operator, our `_or` function will always evaluate x and y (they are passed as arguments) - so we do not have short-circuiting!

In [ ]:
1 or 1/0

1

In [ ]:
_or(1, 1/0)

ZeroDivisionError: division by zero

**and**

`X and Y`: If X is falsy, returns X, otherwise evaluates and returns Y

Once again, note that the truth value of Y is never considered when evaluating `and`, and the `Y` is only evaluated if it needs to be returned (short-circuiting)

In [ ]:
s1 = None
s2 = ''
s3 = 'abc'

In [ ]:
print(s1 and s1[0])
print(s2 and s2[0])
print(s3 and s3[0])

None

a


In [ ]:
print((s1 and s1[0]) or '')
print((s2 and s2[0]) or '')
print((s3 and s3[0]) or '')



a


This technique will also work to return any default value if `s` is an empty string or None:

In [ ]:
print((s1 and s1[0]) or 'n/a')
print((s2 and s2[0]) or 'n/a')
print((s3 and s3[0]) or 'n/a')

n/a
n/a
a


The `not` function

In [ ]:
not 'abc'

False

In [ ]:
not []

True

In [ ]:
bool(None)

False

In [ ]:
not None

True

### 18 - Comparison Operators

**Identity and Membership Operators**

The `is` and `is not` operators will work with any data type since they are comparing the memory addresses of the objects (which are integers)

In [ ]:
0.1 is (3+4j)

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
C:\Users\aserd\AppData\Local\Temp\ipykernel_2072\2697711678.py:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  0.1 is (3+4j)


False

In [ ]:
'a' is [1, 2, 3]

<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
<>:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
C:\Users\aserd\AppData\Local\Temp\ipykernel_2072\3744826144.py:1: SyntaxWarning: "is" with a literal. Did you mean "=="?
  'a' is [1, 2, 3]


False

The `in` and `not in` operators are used with iterables and test membership:

In [ ]:
1 in [1, 2, 3]

True

In [ ]:
[1, 2] in [1, 2, 3]

False

In [ ]:
[1, 2] in [[1,2], [2,3], 'abc']

True

In [ ]:
'key1' in {'key1': 1, 'key2': 2}

True

In [ ]:
1 in {'key1': 1, 'key2': 2}

False

We'll come back to these operators in later sections on iterables and mappings.

**Equality Operators**

The `==` and `!=` operators are value comparision operators.

They will work with mixed types that are comparable in some sense.

For example, you can compare Fraction and Decimal objects, but it would not make sense to compare string and integer objects

In [ ]:
1 == '1'

False

In [ ]:
from decimal import Decimal
from fractions import Fraction

In [ ]:
Decimal('0.1') == Fraction(1, 10)

True

In [ ]:
1 == 1 + 0j

True

In [ ]:
True == Fraction(2, 2)

True

In [ ]:
False == 0j

True

**Ordering Comparisions**

Many, but not all data types have an ordering defined.

For example, complex numbers do not.

In [ ]:
1 + 1j < 2 + 2j

TypeError: '<' not supported between instances of 'complex' and 'complex'

Mixed type ordering comparisons is supported, but again, it needs to make sense:

In [ ]:
1 < 'a'

TypeError: '<' not supported between instances of 'int' and 'str'

In [ ]:
Decimal('0.1') < Fraction(1, 2)

True

**Chained Comparisons**

It is possible to chain comparisons.

For example, in `a<b<c`, Python simply ands the pairwise comparisions: `a<b and b<c`

In [ ]:
1 < 2 < 3

True

In [ ]:
1 < 2 > -5 < 50 > 4

True

In [ ]:
1 < 2 == Decimal('2.0')

True

In [ ]:
import string
'A' < 'a' < 'z' < 'Z' in string.ascii_letters

False